tests/cpp/test_indexing.cpp

```diff
@@ -4407,4 +4420,419 @@ TEST_F(PredicateIndexingTest, UnswitchConsolidationDifferentThreading) {
+  tv1->merge(0, 1);
+
+  TransformPropagator propagator(tv1);
+  MaxLogicalDomainInfoSpanningTree(tv1).traverse(&propagator);
+
+  // Make sure both tv0 and tv1 are indexed without contig indexing
+  struct GetReference : AbstractGetReference {
+    GetReference(const TensorIndexer& indexer, const IdModel& id_model)
+        : AbstractGetReference(indexer, id_model) {}
+
+    Val* getLinearIndex(TensorView* tv, TensorView* maybe_consumer)
+        const override {
+      bool as_consumer = maybe_consumer == nullptr;
+      auto consumer_tv = as_consumer ? tv : maybe_consumer;
+      std::vector<Val*> loop_indices = getLoopIndices(consumer_tv, indexer_);
+
+      auto id0 = divExpr(
+          divExpr(loop_indices.at(0), tv->getLogicalDomain().at(1)->extent()),
+          tv->getLogicalDomain().at(2)->extent());
+      auto id0_extent = mulExpr(
+          tv->getLogicalDomain().at(2)->extent(),
+          tv->getLogicalDomain().at(1)->extent());
+      auto id1 =
+          modExpr(loop_indices.at(0), tv->getLogicalDomain().at(1)->extent());
+      auto id1_extent = tv->getLogicalDomain().at(2)->extent();
+      auto id2 = modExpr(
+          divExpr(loop_indices.at(0), tv->getLogicalDomain().at(1)->extent()),
+          tv->getLogicalDomain().at(2)->extent());
+      return addExpr(
+          addExpr(mulExpr(id0, id0_extent), mulExpr(id1, id1_extent)), id2);
+    }
+  };
+
+  IndexValidator<GetReference>::validate(&fusion, true);
+}
+
+TEST_F(ContigIndexingTest, ConcretizedBroadcastMerge) {
+  Fusion fusion;
+  FusionGuard fg(&fusion);
+
+  // [I0, I1]
+  auto tv0 = makeContigTensor(2);
+  fusion.addInput(tv0);
+  // [I0, I1, I2]
+  auto tv1 = makeContigTensor(3);
+  fusion.addInput(tv1);
+
+  auto tv2 = broadcast(tv0, {false, false, true});
+  auto tv3 = add(tv2, tv1);
+  fusion.addOutput(tv3);
+
+  tv3->merge(1, 2);
+  tv3->merge(0, 1);
+
+  TransformPropagator propagator(tv3);
+  MaxLogicalDomainInfoSpanningTree(tv3).traverse(&propagator);
+
+  tv3->axis(0)->parallelize(ParallelType::TIDx);
+  scheduler_utils::parallelizeAllLike(tv3);
+
+  tv2->setMemoryType(MemoryType::Shared);
+
+  // tv2's broadcast domain is concretized. Previously, this would
+  // have prevented contig indexing.
+
+  struct GetReference : AbstractGetReference {
+    GetReference(const TensorIndexer& indexer, const IdModel& id_model)
+        : AbstractGetReference(indexer, id_model) {}
+
+    Val* getLinearIndex(TensorView* tv, TensorView* maybe_consumer)
+        const override {
+      // Only interested in tv2 here since that's the one that has a
+      // concretized broadcast domain
+      if (tv->name() != 2) {
+        return nullptr;
+      }
+
+      bool as_consumer = maybe_consumer == nullptr;
+      auto consumer_tv = as_consumer ? tv : maybe_consumer;
+      std::vector<Val*> loop_indices = getLoopIndices(consumer_tv, indexer_);
+
+      // When indexed as a consumer, the second merge is a contig
+      // merge, so the index should be just threadIdx.x
```

Comment:
Are you referring to the indexing of `tv2` as a producer? Here's what I got if the merge order is changed (and with `NVFUSER_DISABLE=index_hoist,expr_simplify`):

```
__global__ void CUDAGeneratedKernel(Tensor<float, 2, 2> T0, Tensor<float, 3, 3> T1, Tensor<float, 3, 3> T3) {
  alignas(16) extern __shared__ char array[];
  const unsigned smem_offset = 0;
  float* T2 = reinterpret_cast<float*>(array + smem_offset + 0LL);
  if (((((((((nvfuser_index_t)threadIdx.x) / T0.logical_size[1LL]) >= 0LL) && ((((nvfuser_index_t)threadIdx.x) / T0.logical_size[1LL]) < T0.logical_size[0LL])) && ((((nvfuser_index_t)threadIdx.x) % T0.logical_size[1LL]) >= 0LL)) && ((((nvfuser_index_t)threadIdx.x) % T0.logical_size[1LL]) < T0.logical_size[1LL])) && (((nvfuser_index_t)threadIdx.x) < ((T0.logical_size[0LL] * T0.logical_size[1LL]) * 1LL)))) {
    T2[((nvfuser_index_t)threadIdx.x)]
       = T0[((nvfuser_index_t)threadIdx.x)];
  }
  __syncthreads();
  if ((((((((((((nvfuser_index_t)threadIdx.x) / T1.logical_size[2LL]) / T0.logical_size[1LL]) >= 0LL) && (((((nvfuser_index_t)threadIdx.x) / T1.logical_size[2LL]) / T0.logical_size[1LL]) < T0.logical_size[0LL])) && (((((nvfuser_index_t)threadIdx.x) / T1.logical_size[2LL]) % T0.logical_size[1LL]) >= 0LL)) && (((((nvfuser_index_t)threadIdx.x) / T1.logical_size[2LL]) % T0.logical_size[1LL]) < T0.logical_size[1LL])) && ((((nvfuser_index_t)threadIdx.x) % T1.logical_size[2LL]) >= 0LL)) && ((((nvfuser_index_t)threadIdx.x) % T1.logical_size[2LL]) < T1.logical_size[2LL])) && (((nvfuser_index_t)threadIdx.x) < ((T0.logical_size[0LL] * T0.logical_size[1LL]) * T1.logical_size[2LL])))) {
    T3[((nvfuser_index_t)threadIdx.x)]
      = T2[(((nvfuser_index_t)threadIdx.x) / T1.logical_size[2LL])]
      + T1[((nvfuser_index_t)threadIdx.x)];
  }
}
```

Reply:
Note of course that this is not about this PR per se I am just trying to understand it a bit better.

Anyway, yes you got it; these are the expressions: Original (merging inner to outer):
```c++
(((((nvfuser_index_t)threadIdx.x) /
              (T0.logical_size[1LL] * T1.logical_size[2LL])) *
             T0.logical_size[1LL]) +
            ((((nvfuser_index_t)threadIdx.x) %
              (T0.logical_size[1LL] * T1.logical_size[2LL])) /
             T1.logical_size[2LL]))
```
Merging outer to inner:
```c++
(((nvfuser_index_t)threadIdx.x) / T1.logical_size[2LL])
```
The original index above doesn't get simplified by `simplifyExpr`. I am trying to understand how we might get it to simplify in cases like this. As you said, the path from indexed loops (just the inner loop in this case) to the allocation domain in the ValGraph is through 2 merges in the original, vs one in the reordered case. Since for the purposes of indexing `Merge` is associative, I wonder if this is something we could try and detect automatically. It would probably be a tricky optimization to implement since I think it relies on comparing multiple paths in the ValGraph at once. However, a simple approach would be to always reorder consecutive merges in the path to be outer first. That would naturally shorten the paths for those inner dimensions through the graph and those are the dimensions that are more likely to need indexing.

Reply:
If it's fully contiguous, the order shouldn't matter. Only the output of the last merge would be indexed. This difference, however, happens when a producer has a broadcast domain since it can't just use the merge output of non-broadcast domains. For example:

```
t0 = [i0, b1]
t1 = [i2, i3]
t2 = t0 + t1 // [i4, i5]

flatten all tensors
inlineMost
```

Since `i4*i5` is not exactly mapped with `i0*b1`, the index corresponding to `i4*i5` needs to be decomposed and only the one for `i4` needs to be taken.

So, maybe we could do is to merge broadcast domains only or non-broadcast domains only first. For example:

```
t0 = [b0, b1, i2, i3]
t1 = [i4, i5, i6, i7]
```

If they are merged to `((i0*i1)*(b2*b3))` and `((i4*i5)*(i6*i7))`. Then, initially, we would have an index for `((i4*i5)*(i6*i7))`. We would traverse backward through Merge, to get indices for `(i4*i5)` and `(i6*i7)` and then we would stop as the former index is all we need.

However, if they are merged to `(((i0*i1)*b2)*b3)` and `(((i4*i5)*i6)*i7)`, we would need to traverse Merge twice, first to get an index for `(i4*i5)*i6` and another to get `i4*i5`. 

I'm not sure if there's any significant opportunity. Note that usually a tensor is allocated with its loop domains that are not inlined, so there's even no traversal would be involved.